src/inc/quic_driver_helpers.h

```diff
@@ -0,0 +1,379 @@
+        if (DeviceHandle != INVALID_HANDLE_VALUE) {
+            CloseHandle(DeviceHandle);
+        }
+    }
+    bool Run(
+        _In_ uint32_t IoControlCode,
+        _In_reads_bytes_opt_(InBufferSize)
+            void* InBuffer,
+        _In_ uint32_t InBufferSize,
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        uint32_t Error;
+        OVERLAPPED Overlapped = { 0 };
+        Overlapped.hEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+        if (Overlapped.hEvent == nullptr) {
+            Error = GetLastError();
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "[ lib] ERROR, %u, %s.",
+                Error,
+                "CreateEvent failed");
+            return false;
+        }
+        QuicTraceLogVerbose(
+            TestSendIoctl,
+            "[test] Sending IOCTL %u with %u bytes.",
+            IoGetFunctionCodeFromCtlCode(IoControlCode),
+            InBufferSize);
+        if (!DeviceIoControl(
+                DeviceHandle,
+                IoControlCode,
+                InBuffer, InBufferSize,
+                nullptr, 0,
+                nullptr,
+                &Overlapped)) {
+            Error = GetLastError();
+            if (Error != ERROR_IO_PENDING) {
+                CloseHandle(Overlapped.hEvent);
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "DeviceIoControl failed");
+                return false;
+            }
+        }
+        DWORD dwBytesReturned;
+        if (!GetOverlappedResultEx(
+                DeviceHandle,
+                &Overlapped,
+                &dwBytesReturned,
+                TimeoutMs,
+                FALSE)) {
+            Error = GetLastError();
+            if (Error == WAIT_TIMEOUT) {
+                Error = ERROR_TIMEOUT;
+                CancelIoEx(DeviceHandle, &Overlapped);
+            }
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "[ lib] ERROR, %u, %s.",
+                Error,
+                "GetOverlappedResultEx failed");
+        } else {
+            Error = ERROR_SUCCESS;
+        }
+        CloseHandle(Overlapped.hEvent);
+        return Error == ERROR_SUCCESS;
+    }
+    bool Run(
+        _In_ uint32_t IoControlCode,
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        return Run(IoControlCode, nullptr, 0, TimeoutMs);
+    }
+    template<class T>
+    bool Run(
+        _In_ uint32_t IoControlCode,
+        _In_ const T& Data,
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        return Run(IoControlCode, (void*)&Data, sizeof(Data), TimeoutMs);
+    }
+    bool Read(
+        _In_ uint32_t IoControlCode,
+        _Out_writes_bytes_opt_(OutBufferSize)
+            void* OutBuffer,
+        _In_ uint32_t OutBufferSize,
+        _Out_ LPDWORD OutBufferWritten, // Needs to be LPDWORD to match DeviceIoControl
```

Comment:
Instead of requiring an LPBDWORD, let's make this an optional output uint32_t* parameter. Then, internally use a private DWORD for getting the length; and only set the optional output parameter if specified by the caller.